RA-17-Add-To-Favorites

DenisGubernatorov opened the pull request:
1) Кнопка "Избранное". Перевод на использование setImageResource
2) Реализация функционала кнопки "Избранное". Добавление/удаление из списка с обновлением состояния кнопки

DenisGubernatorov committed:
RA-17-Add-To-Favorites
1) Кнопка "Избранное". Перевод на использование setImageResource

DenisGubernatorov committed:
RA-17-Add-To-Favorites
1) Реализация функционала кнопки "Избранное". Добавление/удаление из списка с обновлением состояния кнопки

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -113,30 +113,55 @@ class RecipeFragment : Fragment() {
+    private fun setFavorites(ids: MutableSet<String>?) {
+        val sharedPrefs =
+            context?.getSharedPreferences(FAVORITES_FILE_KEY, Context.MODE_PRIVATE) ?: return
+        with(sharedPrefs.edit()) {
```
в данном случае with ухудшает читаемость.
SharedPreferences так реализованы, что все функции можно сделать ввиде цепочке
sharedPrefs.edit().
.putStringSet(IDS_KEY, ids)
.commit()

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -113,30 +113,55 @@ class RecipeFragment : Fragment() {
 
         binding.favoritesImage.setOnClickListener {
+
+            val ids = context?.getSharedPreferences(FAVORITES_FILE_KEY, Context.MODE_PRIVATE)
```
у тебя уже есть метод getFavourite. Зачем его дублировать?

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -113,30 +113,55 @@ class RecipeFragment : Fragment() {
 
     private fun setFavoritesButtonImage() {
-        val toDrawId = when (isFavorite) {
+        val contains = getFavorites().contains(recipe?.id.toString())
```
Ты в setOnClick уже проверяешь содержится ли id в списке префов. Запихивай этот результат в boolean переменную и передавай в эту функцию, повторно делать одно и тоже не надо

DenisGubernatorov replied:
Немного перработал логику работы с избранным. Инициализирую соответствующие свойства класса при создании вьюхи. А  то setFavoritesButtonImage вызывается 2 раза. И в слушателе и при инциализации. А если передавать при инициализации, то получалось что нужно было получать список идешек и проверять наличие. А так один раз инициализирую

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -113,30 +113,55 @@ class RecipeFragment : Fragment() {
+    private fun getFavorites(): Set<String> {
+        val sharedPrefs = context?.getSharedPreferences(FAVORITES_FILE_KEY, Context.MODE_PRIVATE)
+        val idsCopy = (sharedPrefs?.getStringSet(IDS_KEY, mutableSetOf()))?.toMutableSet()
+        return idsCopy ?: emptySet()
```
Отдельным пунктом в задании сказано, что надо обернуть возврат в HashSet, там же и объяснена причина

DenisGubernatorov committed:
RA-17-Add-To-Favorites
1) Доработки кода согласно рекомендациям
2) Переработал логику работы с избранным. Состояние иницализируется при создании вьюхи

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -17,6 +17,8 @@ import com.google.android.material.divider.MaterialDividerItemDecoration
 
 class RecipeFragment : Fragment() {
+    private lateinit var ids: java.util.HashSet<String>
+    private lateinit var recipeId: String
```
lateinit не рекомендуется использовать, при обращении до вызова onCreateView, твоё прлиожение упадёт. Лучше просто не давать такую возможность вызова этой перменной. Сделай её локальной для onCreateView и передавай в функции, там где он требуется

DenisGubernatorov replied:
Глаз да глаз за андроид студио надо... Делал их через alt + entr и не посмотрел что он добавил их с latient и для HashSet имя пакета зачем то  добавил

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -17,6 +17,8 @@ import com.google.android.material.divider.MaterialDividerItemDecoration
 import java.io.IOException
 
 class RecipeFragment : Fragment() {
+    private lateinit var ids: java.util.HashSet<String>
```
Лучше укоротить  java.util.HashSet до hashSet

aGroupoid commented on app/src/main/java/com/example/recipesapp/RecipeFragment.kt:
```diff
@@ -17,6 +17,8 @@ import com.google.android.material.divider.MaterialDividerItemDecoration
 class RecipeFragment : Fragment() {
+    private lateinit var ids: java.util.HashSet<String>
+    private lateinit var recipeId: String
     private val recipeImagePrefix = "Property 1="
```
Она всегда констата, вынеси её в componion object как private const val

DenisGubernatorov committed:
RA-17-Add-To-Favorites
1) recipeId и ids сделал локальными в onCreateView и протягиваю их в initUI. Что бы не объявлять их latient.
2) RECIPE_IMAGE_PREFIX вынес в константы.

aGroupoid reviewed (approved):
Молодец, всё верно!
